File: src/widgets/cpp/custom_painting/arrow_box.cpp
#include "arrow_box.h"
#include <QPainter>

ArrowBox::ArrowBox(QWidget *parent) : QWidget(parent) {}

void ArrowBox::paintEvent(QPaintEvent *e) {
  // Draw an arrow pointing downwards

  QPainter painter(this);
  painter.setRenderHint(QPainter::Antialiasing);

  // Draw a straight line
  painter.setPen(QPen(Qt::black, 0.1 * width()));
  painter.drawLine(width() / 2, 0, width() / 2, 0.8 * height());

  // Draw the arrow head
  painter.setPen(QPen(Qt::black, 1));
  painter.setBrush(Qt::black);
  painter.drawPolygon(QPolygonF()
                      << QPointF(width() / 2, height())
                      << QPointF(width() / 2 - 0.2 * width(), 0.8 * height())
                      << QPointF(width() / 2 + 0.2 * width(), 0.8 * height()));
}

QSize ArrowBox::sizeHint() const {
  if (width() < 20 || height() < 20)
    return QSize(20, 20);

  return QSize(width(), height());
}

File: src/widgets/python/custom_painting/arrow_box.py
from PyQt6 import QtWidgets, QtGui, QtCore
from PyQt6.QtCore import Qt
from PyQt6.QtGui import QColor, QPen, QPainter, QPolygonF


class ArrowBox(QtWidgets.QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)

    def sizeHint(self):
        if self.width() < 20 or self.height() < 20:
            return QtCore.QSize(20, 20)
        return QtCore.QSize(self.width(), self.height())

    def paintEvent(self, e):
        painter = QPainter(self)
        painter.setRenderHints(
            QtGui.QPainter.RenderHint.Antialiasing
            | QtGui.QPainter.RenderHint.TextAntialiasing
        )

        pen = QPen(QColor("black"))
        pen.setWidth(0.1 * self.width())
        painter.setPen(pen)
        painter.drawLine(self.width() / 2, 0, self.width() / 2, 0.8 * self.height())
        pen.setWidth(1)
        painter.setBrush(QColor("black"))
        painter.drawPolygon(
            QPolygonF()
            << QtCore.QPointF(self.width() / 2, self.height())
            << QtCore.QPointF(
                self.width() / 2 - 0.2 * self.width(), 0.8 * self.height()
            )
            << QtCore.QPointF(
                self.width() / 2 + 0.2 * self.width(), 0.8 * self.height()
            )
        )
